File: two_pointers/01_valid_palindrome/main.py
def reverse_string(s:str) -> bool:
    newStr = ''
    for c in s:
        if c.isalnum():
            newStr += c.lower()
    return newStr == newStr[::-1]

def two_pointers(s:str) -> bool:
    l, r = 0, len(s) -1

    while l < r:
        while l <r and not alpha_num(s[l]):
            l += 1
        while r > l and not alpha_num(s[r]):
            r -= 1
        if s[l].lower() != s[r].lower():
            return False 
        l, r = l+1,r-1
    
    return True 


def alpha_num(c):
    return (ord('A') <= ord(c) <= ord('Z') or 
                ord('a') <= ord(c) <= ord('z') or 
                ord('0') <= ord(c) <= ord('9'))

File: two_pointers/01_valid_palindrome/Main.java
class Main{

    public boolean reverseString(String s){
        StringBuilder newStr = new StringBuilder();

        for (char c: s.toCharArray()){
            if (Character.isLetterOrDigit(c)){
                newStr.append(Character.toLowerCase(c));
            }
        }
        return newStr.toString().equals(newStr.reverse().toString());
    }

    public boolean twoPointers(String s){
        int l = 0, r=s.length()-1;

        while (l<r){
            while(l<r && !alphaNum(s.charAt(l))){
                l++;
            }
            while(r>l && !alphaNum(s.charAt(r))){
                r--;
            }
            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))){
                return false;
            }
            l++; r--;
        }
        return true;
    }

    public boolean alphaNum(char c){
        return (c >= 'A' && c <= 'Z' || 
        c >= 'a' && c <= 'z' || 
        c >= '0' && c <= '9');
    }


}